File: CSharp/Algorithms/Imperative/Otsu.cs
using System.Runtime.Versioning;

namespace CSharp.Algorithms.Imperative;

[SupportedOSPlatform("windows")]
public class Otsu : IBinarizationAlgorithm
{
    public override void Binarize(byte[] pixels)
    {
        const int nbins = 256;
        var histogram = Histogram(pixels);

        var p = new double[histogram.Length];
        for (var i = 0; i < histogram.Length; i++) p[i] = (double)histogram[i] / pixels.Length;

        var sigmaB = new double[nbins];

        for (var t = 0; t < nbins; t++)
        {
            var qL = 0.0;
            var qH = 0.0;
            for (var i = 0; i < t; i++) qL += p[i];

            for (var i = t; i < nbins; i++) qH += p[i];

            if (qL == 0 || qH == 0) continue;

            var miuL = 0.0;
            var miuH = 0.0;
            for (var i = 0; i < t; i++) miuL += i * p[i];

            for (var i = t; i < nbins; i++) miuH += i * p[i];

            miuH /= qH;
            miuL /= qL;

            sigmaB[t] = qL * qH * (miuL - miuH) * (miuL - miuH);
        }

        var tempThreshold = 0;
        var max = sigmaB[0];
        for (var i = 1; i < sigmaB.Length; i++)
        {
            if (!(sigmaB[i] > max)) continue;
            max = sigmaB[i];
            tempThreshold = i;
        }

        byte threshold = (byte)tempThreshold;

        for (var i = 0; i < pixels.Length; i++)
        {
            pixels[i] = (byte)(pixels[i] > threshold ? 255 : 0);
        }
    }

    private static int[] Histogram(byte[] pixels)
    {
        var histogram = new int[256];
        foreach (var i in pixels) histogram[i]++;
        return histogram;
    }
}

File: FSharp/Algorithms/Otsu.fs
namespace ImageProcessing

module Functional =
    let Histogram2 (pixels: byte[]) : int[] =
        let histogram = Array.zeroCreate<int> 256
        pixels |> Array.iter (fun x -> histogram.[int x] <- histogram.[int x] + 1)
        histogram

    let Histogram (pixels: byte[]) : int[] =
        let mutable histogram = Array.zeroCreate<int> 256
        histogram <- pixels |> Array.fold (fun acc x -> acc.[int x] <- acc.[int x] + 1; acc) histogram
        histogram

    let ThresholdingOtsu2 (pixels: byte[]) : byte =
        let nbins = 256
        let histogram = Histogram pixels

        let p = Array.zeroCreate<double> histogram.Length
        for i = 0 to histogram.Length - 1 do
            p.[i] <- double histogram.[i] / double pixels.Length

        let sigmaB = Array.zeroCreate<double> nbins

        for t = 0 to nbins - 1 do
            let mutable qL = 0.0
            let mutable qH = 0.0
            for i = 0 to t - 1 do
                qL <- qL + p.[i]

            for i = t to nbins - 1 do
                qH <- qH + p.[i]

            if qL = 0.0 || qH = 0.0 then
                ()
            else
                let mutable miuL = 0.0
                let mutable miuH = 0.0
                for i = 0 to t - 1 do
                    miuL <- miuL + (double i * p.[i])

                for i = t to nbins - 1 do
                    miuH <- miuH + (double i * p.[i])

                miuH <- miuH / qH
                miuL <- miuL / qL

                sigmaB.[t] <- qL * qH * (miuL - miuH) * (miuL - miuH)

        let mutable threshold = 0
        let mutable max = sigmaB.[0]
        for i = 1 to sigmaB.Length - 1 do
            if sigmaB.[i] > max then
                max <- sigmaB.[i]
                threshold <- i

        byte threshold

    let ThresholdingOtsu (pixels: byte[]) : byte =
        let nbins = 256
        let histogram = Histogram pixels
        let totalPixels = double pixels.Length

        let probabilities =
            Array.map (fun count -> double count / totalPixels) histogram

        let sigmaB =
            Array.init nbins (fun t ->
                let qL = Array.sum <| Array.sub probabilities 0 t
                let qH = Array.sum <| Array.sub probabilities t (nbins - t)
                if qL = 0.0 || qH = 0.0 then 0.0
                else
                    let miuL = Array.mapi (fun i p -> double i * p) <| Array.sub probabilities 0 t |> Array.sum
                    let miuH = Array.mapi (fun i p -> double (t + i) * p) <| Array.sub probabilities t (nbins - t) |> Array.sum
                    let miuL = miuL / qL
                    let miuH = miuH / qH
                    qL * qH * (miuL - miuH) * (miuL - miuH)
            )

        let threshold = 
            sigmaB |> Array.mapi (fun i v -> i, v) |> Array.maxBy snd |> fst

        byte threshold

    let Binarize (pixels: byte[], threshold: byte) : byte[] =
        pixels |> Array.map (fun x -> if x > threshold then 255uy else 0uy)

    let BinarizeOtsu(pixels: byte[]) =
        let threshold = ThresholdingOtsu pixels
        Binarize (pixels, threshold)